azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,441 @@
+            initial="disconnected",
+            send_event=True,  # Use event_data structures to pass transition arguments
+            finalize_event=_on_transition_complete,
+            queued=True,
+        )
+
+    def register(self, callback=None):
+        """
+        Register the device with the provisioning service.
+        :param:Callback to be called upon finishing the registration process
+        """
+        logger.info("register called from polling machine")
+        self._register_callback = callback
+        self._trig_register()
+
+    def cancel(self, callback=None):
+        """
+        Cancels the current registration process of the device.
+        :param:Callback to be called upon finishing the cancellation process
+        """
+        logger.info("cancel called from polling machine")
+        self._cancel_callback = callback
+        self._trig_cancel()
+
+    def _initialize_register(self, event_data):
+        logger.info("Initializing the registration process.")
+        self._request_response_provider.subscribe(
+            topic=constant.SUBSCRIBE_TOPIC_PROVISIONING, callback=self._on_subscribe_completed
+        )
+
+    def _send_register_request(self, event_data):
+        """
+        Send the registration request.
+        """
+        logger.info("Sending registration request")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+
+        self._operations[rid] = constant.PUBLISH_TOPIC_REGISTRATION.format(rid)
+        self._request_response_provider.send_request(
+            rid=rid,
+            topic=constant.PUBLISH_TOPIC_REGISTRATION.format(rid),
+            request=" ",
+            callback=self._handle_register_response_received,
+        )
+
+    def _query_operation_status(self, event_data):
+        """
+        Poll the service for operation status.
+        """
+        logger.info("Querying operation status from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+        result = event_data.args[0].args[0]
+
+        operation_id = result.operation_id
+        self._operations[rid] = constant.PUBLISH_TOPIC_QUERYING.format(rid, operation_id)
+        self._request_response_provider.send_request(
+            rid=rid,
+            topic=constant.PUBLISH_TOPIC_QUERYING.format(rid, operation_id),
+            request=" ",
+            callback=self._handle_query_response_received,
+        )
+
+    def _handle_register_response_received(self, url_portion, key_values_dict, response):
+        """
+        The function to call in case of a response from a registration request.
+        :param url_portion: The portion of the url containing the status code.
+        :param key_values_dict: The dictionary containing the query parameters of the returned topic.
+        :param response: The complete response from the service.
+        """
+        self._query_timer.cancel()
+
+        url_parts = url_portion.split("/")
```

Comment:
in the iothub library, we have a neat module that abstracts all this stuff out of the application logic called "mqtt_topic". obviously, this is http instead, but the principle still applies.

Reply:
I will wait on this for now to see what kind of changes happen for pipeline.

---
In reply to: [282967152](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282967152) [](ancestors = 282967152)